### ISSUE TITLE
"pub get" consistently fails on Windows

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/pub#1798]
I am working on a Flutter app on Windows and everything used to work fine. However, today I started to see the following error consistently whenever I run `pub get` (even after a fresh restart):
```
Failed to rename directory because access was denied.
This may be caused by a virus scanner or having a file
in the directory open in another application.
```
Yes, I am running a virus scanner on my machine (Windows Defender). I am running on a corporate machine and I don't have the proper rights to turn the scanner off, though.

What do I do now? How do I get past this error to get my Flutter/Dart environment running again on my Windows machine?

**Comment by nex3** [comment]
There's unfortunately not a lot we can do about this in pub. We surface this error because the Windows system call to move the file fails, and only after retrying the call several times in case it's a transient issue.

Note that the issue may well not be a virus scanner, but something else entirely that's taking out a lock on your temp files or your `.pub-cache` directory. Maybe you have some sort of zombie pub process that you need to kill?

**Comment by goderbauer** [comment]
This also happens after a fresh restart of the system, so any zombie pub processes are unlikely.

**Comment by goderbauer** [comment]
Does pub have to move the files? Could it copy the files to a new location and then attempt to delete the old version? If that fails, it could just print a warning and then continues on with its work. That way, I would end up with a working environment - with possibly some junk left on my hard disk, which I could delete manually (or if at all possible the junk could be left behind in the system's temp folder and be cleaned up by the OS at an appropriate time).

Or is copying of the locked files also disallowed?

**Comment by gspencergoog** [comment]
This is for moving a new package in place in the cache: the case where it's downloaded a new package and wants to move the old one out of the way to replace it, and can't rename the old one, right?

It seems to me that an rsync-like process might work where it copies files that have changed, creates directories that have been created, and removes directories that have been removed, and where it was blocked by the system from deleting a file or empty directory, that could be added to a queue, and retried a few times after all the successful operations have finished.  This would make it so that any virus scanner or indexer would have a chance to be done with the file so it can be deleted.  Right now, it has to wait until the scanner is done with the entire directory in order to succeed, and it doesn't do any retries (that I know of). Of course, failure could leave you in a broken state, but there are probably ways to fix that (make a copy of the destination before copying, for instance).

A first step would be to just retry a number of times to rename the directory, with a delay between each retry.

**Comment by gspencergoog** [comment]
This basically makes pub (and flutter) unusable for any machine with a virus scanner, which, as this is Windows, is a pretty common case.  Any chance there are some other solutions we can try?  For instance, you can at least tell the user what process/executable is holding the lock, using [something like this](https://stackoverflow.com/questions/317071/how-do-i-find-out-which-process-is-locking-a-file-using-net)?

Surely we're not the first to run into locking problems on Windows.  There must be a better solution than "turn off your virus scanner, reboot, and cross your fingers".  Many corporate users don't have the option to turn off their virus scanners (although rebooting and crossing fingers is usually still allowed).

**Comment by nex3** [comment]
> Does pub have to move the files? Could it copy the files to a new location and then attempt to delete the old version?

It's important that we do the move atomically. Otherwise, if we encounter transient OS issues like this, the cache can end up in an inconsistent state that will at best force the user to run `pub cache repair` and at worst break pub in unexpected ways.

> This basically makes pub (and flutter) unusable for any machine with a virus scanner, which, as this is Windows, is a pretty common case.

It seems to be extremely rare that users run into this in practice. This is the first report I've seen of it since we started retrying on Windows when the move failed, which is why I think something unusual is going on in @goderbauer's system. I wouldn't expect a normal virus scanner to hold a lock long enough for this to fail so consistently.

> For instance, you can at least tell the user what process/executable is holding the lock, using [something like this](https://stackoverflow.com/questions/317071/how-do-i-find-out-which-process-is-locking-a-file-using-net)?

If `dart:io` provided access to this information, we could definitely surface it.

**Comment by goderbauer** [comment]
> It's important that we do the move atomically. Otherwise, if we encounter transient OS issues like this, the cache can end up in an inconsistent state that will at best force the user to run pub cache repair and at worst break pub in unexpected ways.

Can you explain this? How is a copy in this case less atomic than a move? Currently, pub also leaves trash around in the temp directory if it can't do the move. Wouldn't a copy followed by a (possibly failing) delete operation just run the same risk of inconstant state?

> which is why I think something unusual is going on in @goderbauer's system.

What can this "something unusual" be? Happy to provide more information about my machine.

**Comment by nex3** [comment]
> Can you explain this? How is a copy in this case less atomic than a move?

Copying is implemented by the OS by individually copying each component file of a directory, while moving is atomic at the filesystem level. If a move fails, there's guaranteed to be nothing in the target location, which isn't the case for a copy.

> Currently, pub also leaves trash around in the temp directory if it can't do the move. Wouldn't a copy followed by a (possibly failing) delete operation just run the same risk of inconstant state?

Leaving junk in the temp directory isn't a big deal, since future pub processes won't read from there. What we really care about is leaving junk in the pub cache. For example, if we create a directory there that looks like a package but is missing some of its files, users' programs will fail and even `pub cache repair` might not notice that it's corrupted.

> What can this "something unusual" be? Happy to provide more information about my machine.

I'm not enough of a Windows expert to say for sure. I think your best bet is running one of the diagnostic tools @gspencergoog linked to to see exactly what process is holding the lock.

**Comment by Hixie** [comment]
FWIW, we also have a report that antivirus on Windows causes problems here: https://github.com/flutter/flutter/issues/12341
We consider that bug a high priority for our current milestone.

**Comment by matanlurey** [comment]
Hi @nex3 and interested Dartisan friends.

Can you schedule a VC with the Flutter folks and see if we can resolve this together?

**Comment by nex3** [comment]
If there are any concrete proposals for addressing this while preserving atomicity, I'm happy to work on them. But as far as I can tell, this is essentially an operating system issue, and I don't see a way to work around it in pub beyond what we're already doing.

**Comment by goderbauer** [comment]
Have you investigated what other packages managers do on Windows to get around this?

**Comment by natebosch** [comment]
> while preserving atomicity

What if we look for a pattern that doesn't require atomicity? For instance we could drop a file in each package that says "this directory is complete" or keep a manifest somewhere. As long as we don't write that file until the non-atomic copy is complete, and don't reference incomplete packages in `.packages` files, we should be OK.

It might be challenging, though, to upgrade from a pub cache without this marker to one that has it...

**Comment by natebosch** [comment]
For anyone able to repro this with a windows machine:

You should be able to use [process explorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) and the "Handle" menu to track down the process that is holding on to the file handle.

@nex3 - do we currently print out the path of the directory that caused the issue? If not, can we prioritize adding that?

**Comment by goderbauer** [comment]
@natebosch The directory with the lock is printed out. I was not able to catch which process holds on to the lock with process explorer. Seems like process explorer takes a few seconds to find all handles and by that time the lock was released already.

Having said that, I am almost certain that the lock is held by Windows Defender, the build-in virus scanner. If I move my pub cache to a directory excluded by the scanner I can't reproduce the failure anymore.

**Comment by nex3** [comment]
If the handle is released within a few seconds, maybe the solution is to increase the time between retries.

**Comment by asrendon** [comment]
I thought at first the issue might have been with 'git' and github as i received a notice once I initialized my project directory, but didn't write down the notice.

In any rate, this error was encountered in AndroidStudio. So I closed down Android studio, and manually ran the "flutter package get" command in an elevated command prompt once I was in the project directory and then everything was resolved.
It does suck because some of us use this in an enterprise and cannot work around virus scanners for our dev environment.. Talk to our IT departments :) I know all too well as I did helpdesk for 2 years.

**Comment by cmarques2** [comment]
Just wanted to add some info for tracking this down.

My fail (can't rename dir) is always happening when processing sass 1.18.0. If I put in a dependency override to 1.17.3 it all works. Maybe that zip file has a bit pattern that matches a known virus. 

Downloading sass 1.18.0 on another machine and copying it into pub.dartlang.org is my current workaround.

**Comment by vivekanandasuggu** [comment]
Even for the flutter web when I run the command  "flutter pub global activate webdev''  got the following error-
Failed to rename directory because access was denied.
This may be caused by a virus scanner or having a file
in the directory open in another application.
pub failed (1)

**Comment by EndersonPro** [comment]
> ```
> Failed to rename directory because access was denied.
> This may be caused by a virus scanner or having a file
> in the directory open in another application.
> ```
The solution I found for this problem was to delete the cache folder that is in the `flutter/bin/cache` directory, and reinstall dart-sdk.

**Comment by humazed** [comment]
what worked for me is to delete `.dart_tool`

**Comment by ronnieoverby** [comment]
Where is the directory in question? Where is the pub cache located on windows?

**Comment by jonasfj** [comment]
> Where is the directory in question? Where is the pub cache located on windows?

`%APPDATA%\Pub\Cache\`, see:

https://github.com/dart-lang/pub/blob/df0f72daaa724e29ed6075e0fb5549a6d6dc5daf/lib/src/system_cache.dart#L38-L39

**Comment by zhengzongsheng** [comment]
![image](https://user-images.githubusercontent.com/34943180/62830568-e1e19980-bc43-11e9-89af-8aafc1ef990d.png)
Run your compiler this way

**Comment by robinbakkerus** [comment]
I also frequently encountered the following error while running
> flutter pub get (or upgrade)
Failed to rename directory because access was denied.

This problem kept occurring also after a completely fresh install and after deleting the 'Pub' folder under AppData.

But what I noticed is that the error did NOT occur anymore after I disconnected from my 'Pulse secure' session !

**Comment by jonasfj** [comment]
> Failed to rename directory because access was denied.

This could also be some I/O issue on the machine.

**Comment by sbhaya** [comment]
![Capture](https://user-images.githubusercontent.com/35295462/67013864-728b9880-f111-11e9-9da7-fdc38dc6e82c.PNG)
[https://flutter.dev/docs/get-started/install/windows](url)
Installation bundle of the latest stable release of the Flutter SDK 
flutter_windows_v1.9.1+hotfix.4-stable.zip helped me resolve this issue.

**Comment by shrkamat** [comment]
Used latest flutter (hotfix.5), still face the same issue.
```
flutter doctor -v
[√] Flutter (Channel stable, v1.9.1+hotfix.5, on Microsoft Windows [Version 10.0.17134.1069], locale en-IN)
    • Flutter version 1.9.1+hotfix.5 at C:\skm\engine\flutter
    • Framework revision 1aedbb1835 (26 hours ago), 2019-10-17 08:37:27 -0700
    • Engine revision b863200c37
    • Dart version 2.5.0

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\shrkamat\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 40.2.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] IntelliJ IDEA Community Edition (version 2018.3)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2018.3.4
    • Flutter plugin version 33.3.2
    • Dart plugin version 183.5901

[!] VS Code (version 1.33.0)
    • VS Code at C:\Users\shrkamat\AppData\Local\Programs\Microsoft VS Code
    X Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 9 (API 28) (emulator)

! Doctor found issues in 1 category.
```

**Comment by YashawantParab** [comment]
Getting Same error while saving and upgrading pubspec.yaml file.
Error: 
Failed to rename directory because access was denied.
This may be caused by a virus scanner or having a file
in the directory open in another application.
Running "flutter pub get" in <projectName>...
pub get failed (1)

Solution:
Just Disable Antivirus which installed in your system for 15mins and save & get packages. This helped me to resolve this issue.

**Comment by fringefilmsoz** [comment]
I had the same error. 
Restarting Windows sometimes unlocks the file, but not always.
I was able to run the pub get immediately after a reboot, using Command Prompt.

**Comment by basemnoori1990** [comment]
if you have avast just uninstall it .

**Comment by godilite** [comment]
Here is what I found, this fill `C:\flutter\bin\cache\dart-sdk\lib\_internal\pub\asset\7zip\7za.exe` was actually duplicated in the folder by App deployment service in windows 10, I deleted the folder `C:\flutter\bin\cache\dart-sdk\lib\_internal\pub\asset\7zip\` and replaced it with a fresh copy rather than deleting the whole of `C:\flutter\bin\cache\` and it worked.

**Comment by DavidBruchmann** [comment]
I had the same issue and just repeated the installation, everything was working then.
Virus Scanners I never did switch off and I doubt that it has to do with it in my case.

**Comment by BobbyBriQz** [comment]
I just resolved this issue on my system, here's how. I closed down the AppX deployment service with Task Manager, I deleted the cache folder at <flutter-install-location>\bin\cache. Using winrar I extracted only the cache folder from the flutter zip file downloaded during setup and replaced the deleted cache folder. I called "flutter clean" just to be safe, and then called "flutter pub get" and it worked fine.

**Comment by blackcrabb** [comment]
I solved it by re-cloning flutter.git and changing the environment variable to the new path and relocating the SDK path.
I found only this way easier to get away from the infinite loop.

**Comment by rahulserver** [comment]
I solved it by deleting the build folder and .dart_tool folders. And then ran pub get again. This worked for me.

**Comment by saurabhrayakwar** [comment]
`pub cache repair` did the trick for me

**Comment by chethandhruva** [comment]
Hey guys, I ran into the same error, and here's what I did to fix it.

run chmod 777 /c/path/flutter/bin/cache/dart-sdk/bin in git bash
and try running pub get in git bash as an administrator.
It worked for me.

Similarly we can try flutter doctor as well after running the above command

**Comment by Pmmlabs** [comment]
One more solution: **in case you can't disable antivirus**, you can run `pub cache repair`, it will print packages that can't be moved, then go to `flutter\.pub-cache\_temp` and rename every `dir***` to name from "repair" output (use pubspec file inside each directory to get right one). After this, move these directories to `flutter\.pub-cache\hosted\pub.dartlang.org` and run `flutter pub get` again. Works for me.

**Comment by Miiite** [comment]
We have windows machine in my company with the same issue. We also have mcAfee installed on these machines.
`flutter pub get` fails every time on my machine and the computer of my colleague.
BUT if we run `flutter pub get --verbose` ... it works on both our machines. Why ? I have no clue.

**Comment by harnarayans** [comment]
> We have windows machine in my company with the same issue. We also have mcAfee installed on these machines.
> `flutter pub get` fails every time on my machine and the computer of my colleague.
> BUT if we run `flutter pub get --verbose` ... it works on both our machines. Why ? I have no clue.

Same problem with me. `flutter pub get --verbose`  works fine and after that even `flutter bub get` works fine.

**Comment by Varun11940** [comment]
> > while preserving atomicity
> 
> What if we look for a pattern that doesn't require atomicity? For instance we could drop a file in each package that says "this directory is complete" or keep a manifest somewhere. As long as we don't write that file until the non-atomic copy is complete, and don't reference incomplete packages in `.packages` files, we should be OK.
> 
> It might be challenging, though, to upgrade from a pub cache without this marker to one that has it...



> pub cache repair

whenever I am trying to run pub cache repair it is showing like this-"pub : The term 'pub' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ pub cache repair
+ ~~~
    + CategoryInfo          : ObjectNotFound: (pub:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException"

**Comment by rdxking** [comment]
> 
> 
> I am working on a Flutter app on Windows and everything used to work fine. However, today I started to see the following error consistently whenever I run `pub get` (even after a fresh restart):
> 
> ```
> Failed to rename directory because access was denied.
> This may be caused by a virus scanner or having a file
> in the directory open in another application.
> ```
> 
> Yes, I am running a virus scanner on my machine (Windows Defender). I am running on a corporate machine and I don't have the proper rights to turn the scanner off, though.
> 
> What do I do now? How do I get past this error to get my Flutter/Dart environment running again on my Windows machine?

Hello dear,

There is a solution, as it works for me. just reboot/restart you system. and open your android studio once again and then you will see the message at right side of your bottom for pub get command just click on there and everything will be works fine.

please note: don't open any directory as I have just opened my android studio instantly after rebooting, and just got that message at right side of bottom for pub get.


Hope this will helps you.

**Comment by alenros** [comment]
Other solutions didn't work for me, but deleting `pubspec.lock` solved it.

**Comment by Varun11940** [comment]
Thank you problem is solved

**Comment by Akshay-04** [comment]
replace your flutter folder with a new flutter folder at the same spot one.meaning  I reinstalled  flutter and my issue was solved.

**Comment by Varun11940** [comment]
Thank you issue is fixed

**Comment by atapas** [comment]
I'm still facing it. I tried reinstalling the flutter, recreating the project, etc. I am still facing it. Unfortunately, I can not uninstall/disable antivirus as my org policy doesn't allow me. At this point, no clue what to try next.

**Comment by atapas** [comment]
> I'm still facing it. I tried reinstalling the flutter, recreating the project, etc. I am still facing it. Unfortunately, I can not uninstall/disable antivirus as my org policy doesn't allow me. At this point, no clue what to try next.

# I found this as a solution

When there was a dead-end, I found this as a solution. Nothing else worked for me so far. This one just worked like a charm. Such a relief!

https://stackoverflow.com/a/60717742/1127192

**Comment by urajarama** [comment]
Hi All,

 I had similar issue while using shared_preferences 2.0.5 in dependency. Followed below steps and it got fixed.

- Downloaded the package from https://pub.dev/packages/shared_preferences/versions
- Extracted the folder and pasted in C:\<<path>>\flutter_windows_2.2.0-stable\flutter\packages
- Updated .packages file in my flutter project with shared_preferences:file:///C:<<path>>/flutter_windows_2.2.0-stable/flutter/packages/shared_preferences-2.0.5/lib/
- run the command again **flutter pub get**

It worked for me

**Comment by jdavidzapatab** [comment]
For those having this issue on Windows, an alternative would be using Docker. Run this command from the root directory of your Dart Project:

`docker run -it --rm --name dartvm -v "%CD%":/app -w /app dart:stable /bin/bash`

From there, just run the needed dart commands as needed:

```
dart pub get
dart analyze
dart test
dart run bin/my_app.dart
```

**Comment by AkhilSharma90** [comment]
guys - quick fix - worked for me - delete cache file in flutter/bin/cache, then use powershell as admin, switch to flutter dev channel and flutter upgrade - flutter will upgrade and also download the latest dart version that's compatible.

**Comment by priyadarshu** [comment]
> guys - quick fix - worked for me - delete cache file in flutter/bin/cache, then use powershell as admin, switch to flutter dev channel and flutter upgrade - flutter will upgrade and also download the latest dart version that's compatible.

Thanks a lot @AkhilSharma90 it worked for me!

**Comment by mmfarahat** [comment]
I had the same issue and running Vscode as administrator solved my problem.

**Comment by Mohitmadhav** [comment]
Since I had the setup, I deleted the Flutter SDK, disabled my antivirus, and reinstalled it but this time, in a different directory like G:/flutter/bin.
It might have to do with some permission issues. But worked for me anyways.

**Comment by FelixWielinga** [comment]
I tried almost everything mentioned above, including delete flutter and reinstall, same for sdk. Even made a new project and cloned from git again. Still pub get kept throwing the error (exitcode 1;...)

then I did pub upgrade and after that pub get did work!

**Comment by FelixWielinga** [comment]
I tried almost everything mentioned above, including delete flutter and reinstall, same for sdk. Even made a new project and cloned from git again. Still pub get kept throwing the error (exitcode 1;...)

then I did pub upgrade and after that pub get did work!

**Comment by ArtemAlt** [comment]
> > I'm still facing it. I tried reinstalling the flutter, recreating the project, etc. I am still facing it. Unfortunately, I can not uninstall/disable antivirus as my org policy doesn't allow me. At this point, no clue what to try next.
> 
> # I found this as a solution
> When there was a dead-end, I found this as a solution. Nothing else worked for me so far. This one just worked like a charm. Such a relief!
> 
> https://stackoverflow.com/a/60717742/1127192
> 
> I've provided the `allow` permission for both `bin` and `.pub-cache` folder. This screenshot may be useful,
> 
> ![image](https://user-images.githubusercontent.com/3633137/119250790-1972c900-bbc0-11eb-8140-831511b10de2.png)

nice!! working windows 10  and flutter_windows_2.5.0-stable

**Comment by topcheese** [comment]
First it was only happening with one package for me, so what I had to do was run windows PowerShell in admin mode.  It was the ObjectBox package that had the problem.

**Comment by fidelisakilan** [comment]
> Other solutions didn't work for me, but deleting `pubspec.lock` solved it.

This solution works for me, Thanks

**Comment by anishre** [comment]
is anyone using the audioplayers package by chance and then getting this error?

**Comment by LaoTie0815** [comment]
I have solved it. I recommend instead of C:/src/flutter, you put the flutter folder inside C:/ direct. Would be like this:

C:/flutter

If that doesn't solve it, try pausing your anti-virus just for us to test. It could be that it is blocking the permissions on local disk C folder. It's not very common, but it can happen sometimes.

**Comment by Varun11940** [comment]
Ok thank you my issue is resolved

**Comment by sigurdm** [comment]
We have discussed two possible workarounds/solutions here:

1.: increase the number+timing of retries:
  * This would be very easy to do, and might be enough. We need a way to confirm this though.

2.: unpack cache files in place, instead of relying on atomic rename.
  * This seems to be what npm has ended up doing.
  * We should take care to always unpack pubspec.yaml last
    - in that way we can consider a package installed depending on the existence of `pubspec.yaml`. 
    - We should ensure all code respects this invariant.
    - We should ensure that unfinished packages are deleted by `cache clean`.

**Comment by sigurdm** [comment]
We have done (1) in https://github.com/dart-lang/pub/pull/3451, extended the number of retries.

Closing tentatively. Please try it out in dart 2.18 and open a new issue if it still reproduces.
